crates/extensions/c8y_auth_proxy/src/tokens.rs

```diff
@@ -0,0 +1,43 @@
+use std::sync::Arc;
+
+use c8y_http_proxy::credentials::JwtRetriever;
+use tokio::sync::Mutex;
+
+#[derive(Clone)]
+pub struct SharedTokenManager(Arc<Mutex<TokenManager>>);
+
+impl SharedTokenManager {
+    pub async fn not_matching(&self, input: Option<&str>) -> Arc<str> {
+        self.0.lock().await.not_matching(input).await
+    }
+}
+
+pub struct TokenManager {
+    recv: JwtRetriever,
+    cached: Option<Arc<str>>,
+}
+
+impl TokenManager {
+    pub fn new(recv: JwtRetriever) -> Self {
+        Self { recv, cached: None }
+    }
+
+    pub fn shared(self) -> SharedTokenManager {
+        SharedTokenManager(Arc::new(Mutex::new(self)))
+    }
+}
+
+impl TokenManager {
+    async fn not_matching(&mut self, input: Option<&str>) -> Arc<str> {
```

Comment:
What is done by this method is mysterious.

Provided with the cached token, it returns a fresh one.
And provided with a silly token, it returns the cached one.

What is the point?

Reply:
I think I was trying to be a bit unnecessarily clever here. The goal was that if we receive a request and that request fails, we can call this method with the token that failed, and if another request has caused the token to be regenerated in the meantime, we will just return the new cached token rather than making a new JWT request.

Reply:
I got it!

Indeed unnecessary clever. But it's okay. Just put you full response as a comment: "The goal is ...".